core/admin/mailu/models.py

```diff
@@ -493,10 +507,18 @@ class Token(Base):
     ip = db.Column(db.String(255))
 
     def check_password(self, password):
-        return hash.sha256_crypt.verify(password, self.password)
+        if self.password.startswith("$5$"):
+            if hash.sha256_crypt.verify(password, self.password):
+                self.set_password(password)
+                db.session.add(self)
+                db.session.commit()
+                return True
+            return False
+        return hash.pbkdf2_sha256.verify(password, self.password)
```

Comment:
Is there a particular reason why the tokens have a hash detection on their own instead of relying on passlib?

Reply:
Simplicity. It's a lot more verbose to use a context... and unlike for passwords we are 100% sure of what we are working with here.

If you feel strongly about it I'll change it, but IMHO it's fine and explicit like this

Reply:
I'm implementing a config import/export (json/yaml) using marshmallow and I would need to verify the imported hash is using the correct algorithm if this is implemented without passlib's autodetection.
So my first thought is: check the token like the user password. 
But there are other problems arising when implementing tokens in this fixed way. For example I am unable to make sure the hashed token has the correct length.
Perhaps this just has to be documented somewhere so anyone mass-creating and importing tokens via config-import does not use tokens with length 64...
Alex

Reply:
Just to clarify: the code it replaces wasn't doing any validation nor auto-detection either... 

This is mixing two different problems: import/export is fine... The question is whether mailu should add complexity to allow 3rd party tools to modify what has been imported/exported (and/or standardize what the blob is made of). 

IMHO the answer to that question is "no". 3rd party tools should use the API to get mailu to generate token it likes/can work with... This enables for simplifications and optimizations to be made (like avoiding the token checking code altogether if the tokens aren't in the desired format).

Reply:
You are correct - the user_import function for example is not really checking its input now.
I also think anyone fiddling around with the exported yaml/json should know what he/she is doing.
As you said: when using the admin interface or api everything will be fine (as mailu is controlling the creation).

It's okay for me to leave the model the way it is (not too much verification) and verify user input on first touch (in multiple places: cmdline, api, ...) - I'll try not to duplicate too much code.

Reply:
Seems it's not only less verbose, but constructing a context every time also takes ~three times as long on my notebook. Considering that there will be many tokens, but only one proper password I think it makes sense to diverge here.

If it's fine for you we can resolve this conversation.